src/host/eigensolver.cpp

```diff
@@ -18,99 +19,119 @@
 
 namespace bipp {
 namespace host {
+
 template <typename T>
-static auto copy_2d(std::size_t m, std::size_t n, const T* a, std::size_t lda, T* b,
-                    std::size_t ldb) {
-  if (lda == ldb && lda == m) {
-    std::memcpy(b, a, m * n * sizeof(T));
+static auto copy_lower_triangle_at_indices(std::size_t m, const std::vector<std::size_t>& indices,
+                                           const T* a, std::size_t lda, T* b, std::size_t ldb) {
+  const std::size_t mReduced = indices.size();
+  if (mReduced == m) {
+    for (std::size_t col = 0; col < mReduced; ++col) {
+      std::memcpy(b + col * ldb + col, a + col * lda + col, (mReduced - col) * sizeof(T));
+    }
   } else {
-    for (std::size_t i = 0; i < n; ++i) {
-      std::memcpy(b + i * ldb, a + i * lda, m * sizeof(T));
+    for (std::size_t col = 0; col < mReduced; ++col) {
+      const auto colIdx = indices[col];
+      for (std::size_t row = col; row < mReduced; ++row) {
+        const auto rowIdx = indices[row];
+        b[col * mReduced + row] = a[colIdx * lda + rowIdx];
+      }
     }
   }
 }
 
 template <typename T>
 auto eigh(ContextInternal& ctx, std::size_t m, std::size_t nEig, const std::complex<T>* a,
-          std::size_t lda, const std::complex<T>* b, std::size_t ldb, std::size_t* nEigOut, T* d,
-          std::complex<T>* v, std::size_t ldv) -> void {
-  // copy input into buffer since eigensolver will overwrite
+          std::size_t lda, const std::complex<T>* b, std::size_t ldb, T* d, std::complex<T>* v,
+          std::size_t ldv) -> void {
   auto bufferA = Buffer<std::complex<T>>(ctx.host_alloc(), m * m);
-  copy_2d(m, m, a, lda, bufferA.get(), m);
-
   auto bufferV = Buffer<std::complex<T>>(ctx.host_alloc(), m * m);
   auto bufferD = Buffer<T>(ctx.host_alloc(), m);
   auto bufferIfail = Buffer<int>(ctx.host_alloc(), m);
+  std::complex<T>* aReduced = bufferA.get();
 
-  auto bufferPtrD = bufferD.get();
-  auto bufferPtrV = bufferV.get();
+  std::vector<std::size_t> indices;
+  indices.reserve(m);
 
-  int hMeig = 0;
-  if (lapack::eigh_solve(LapackeLayout::COL_MAJOR, 'V', 'V', 'L', m, bufferA.get(), m,
-                         std::numeric_limits<T>::epsilon(), std::numeric_limits<T>::max(), 0, m - 1,
-                         &hMeig, bufferD.get(), bufferV.get(), m, bufferIfail.get())) {
-    throw EigensolverError();
+  // find indices of working coloumns
```

Comment:
I think the entire column should be considered for the evaluation of the sum, not only elements from the triangular part of the matrix. With the visibility matrix this would always discard the last column as the diagonal elements are zero.

Reply:
The diagonal is not usually zero from what I've seen. The eigensolver functions only access the lower triangle, since the matrix is hermitian. So ideally one would instead compute the sum of a column and the complex conjugate of a row within the lower triangle. But so far it's seems to work as is, so I'm not sure if that effort would be worth it.

Reply:
I've now added a combined check of row and coloumn within the lower triangle, making use of the hermitian property of the matrix.